module/zfs/dbuf.c

```diff
@@ -1027,6 +1027,20 @@ dbuf_read_impl(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags)
 		 */
 		int bonuslen = MIN(dn->dn_bonuslen, dn->dn_phys->dn_bonuslen);
 		int max_bonuslen = DN_SLOTS_TO_BONUSLEN(dn->dn_num_slots);
+		arc_buf_t *dn_buf = (dn->dn_dbuf != NULL) ?
+		    dn->dn_dbuf->db_buf : NULL;
+
+		if (dn_buf != NULL && arc_is_encrypted(dn_buf) &&
+		    DMU_OT_IS_ENCRYPTED(dn->dn_bonustype) &&
+		    !(flags & DB_RF_NO_DECRYPT)) {
+			err = arc_untransform(dn_buf, dn->dn_objset->os_spa,
+			    dmu_objset_id(dn->dn_objset), B_TRUE);
```

Comment:
I could probably lock the header. This code would still look racy though, as you pointed out. I guess I can just leave a comment about it.